### PR TITLE
feat: [DHIS2-18250] Breadcrumb for event & enrollment pages

### PR DESCRIPTION
**Pull request by eirikhaugstulen**
Tech-summary:
Added two new components that helps the user navigate our app

- EnrollmentBreadcrumb (Used on all enrollment pages
    - {programStageWorkingListName} / {workingListName} / {TET} list / {Search}
    - Enrollment dashboard
    - View event (if applicable)
    - Edit event (if applicable)
- EventBreadcrumb (Old view and edit event pages)
    - Event list
    - View event
    - Edit event (if applicable)

### CODE REVIEW

File: src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.component.js [hunk 1]
```diff
@@ -73,6 +76,11 @@ export const EnrollmentEditEventPageComponent = ({
             pageLayout={pageLayout}
             currentPage={mode === EnrollmentPageKeys.EDIT_EVENT ? EnrollmentPageKeys.EDIT_EVENT : EnrollmentPageKeys.VIEW_EVENT}
             availableWidgets={WidgetsForEnrollmentEventEdit}
+            userInteractionInProgress={mode === EnrollmentPageKeys.EDIT_EVENT}
```
Comment (eirikhaugstulen) [inline src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.component.js:79]: We should probably refactor this at some point, but just following the logic in `src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/TopBar.container.js`

File: src/core_modules/capture-core/components/Breadcrumbs/EventBreadcrumb/EventBreadcrumb.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                        classes={classes}
+                    />
+                    {index < (breadcrumbItems.length - 1) && (
+                        <IconChevronRight16 color={colors.grey800} />
+                    )}
+                </React.Fragment>
+            ))}
+
+            <DiscardDialog
+                open={openWarning === 'mainPage'}
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/Breadcrumbs/EventBreadcrumb/EventBreadcrumb.js:95]: open={openWarning === pageKeys.MAIN_PAGE}

File: src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/hooks/useWorkingListLabel.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        }, {
+            enabled: !!selectedTemplateId &&
+                !DefaultFilterLabels[selectedTemplateId] &&
+                !isDefaultTemplate &&
+                isSameProgram,
+            onError: () => {
+                setShouldFetchPSFilter(true);
+            },
+        },
+    );
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/hooks/useWorkingListLabel.js:67]: I don't think it is necessary to call the API to get the WL names. If selectedTemplateId is stored in Redux, its name is also stored in Redux in the workingListsTemplates.teiList.templates array. Can you check it?

File: src/core_modules/capture-core/components/Breadcrumbs/EventBreadcrumb/hooks/useWorkingListLabel.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            resource: 'eventFilters',
+            id: selectedTemplateId,
+            params: {
+                fields: 'id,displayName',
+            },
+        },
+        {
+            enabled: !loading && !!selectedTemplateId && !isDefaultTemplate && isSameProgram,
+        },
+    );
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/Breadcrumbs/EventBreadcrumb/hooks/useWorkingListLabel.js:34]: Same concern as mentioned here. If selectedTemplateId is stored in Redux, its name is also stored in Redux in the workingListsTemplates.eventList.templates array.

File: src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/EnrollmentBreadcrumb.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            condition: page === EnrollmentPageKeys.EDIT_EVENT,
+        },
+        {
+            key: 'newEvent',
+            onClick: () => {},
+            label: i18n.t('New event'),
+            selected: page === EnrollmentPageKeys.NEW_EVENT,
+            condition: page === EnrollmentPageKeys.NEW_EVENT,
+        },
+    ].filter(item => item.condition !== false)), [
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/EnrollmentBreadcrumb.js:115]: Makes sense! Good catch!

File: src/core_modules/capture-core/components/Breadcrumbs/common/BreadcrumbItem.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
+// @flow
+import React, { type ComponentType } from 'react';
+import cx from 'classnames';
+import { withStyles } from '@material-ui/core/styles';
+import { colors } from '@dhis2/ui';
+
+type Props = {
+    label: string,
+    onClick: () => void,
+    selected: boolean,
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Breadcrumbs/common/BreadcrumbItem.js:10]: The `classes` prop should be automatically added by `ComponentType<$Diff<Props, CssClasses>>` 👍

File: src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/EnrollmentBreadcrumb.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                handleNavigation(onBackToViewEvent, 'viewEvent');
+            },
+            label: i18n.t('View event'),
+            selected: page === EnrollmentPageKeys.VIEW_EVENT,
+            condition: page === EnrollmentPageKeys.VIEW_EVENT ||
+                (page === EnrollmentPageKeys.EDIT_EVENT && !eventIsScheduled(eventStatus)),
+        },
+        {
+            key: 'editEvent',
+            onClick: () => {},
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/EnrollmentBreadcrumb.js:103]: Removing empty attributes usually makes sense, but in this case I've added an empty callback for two main reasons:
1. Flow - flow does not like to infer optional attributes on an object. This means I would have to write separate types for objects with or without this onClick. 
2. Removing the callback would mean we would have to do further validation to check if onClick really was a function further down the tree.

In this case, I  think it makes sense to just keep an empty callback to bypass both of them. Do you agree?

**Comment by eirikhaugstulen** [comment]
Thanks for the review @simonadomnisoru & @henrikmv! Should be sorted now 🚀

**Review by simonadomnisoru — APPROVED** [review]
> Thanks for the review @simonadomnisoru & @henrikmv! Should be sorted now 🚀

Looks great now 👏 !

File: src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/hooks/useWorkingListLabel.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    cancelled: i18n.t('Cancelled enrollments'),
+};
+
+export const useWorkingListLabel = ({
+    programId,
+    trackedEntityName,
+    displayFrontPageList,
+}: Props) => {
+    const workingListTemplates = useSelector(({ workingListsTemplates }) => workingListsTemplates?.teiList);
+    const workingListProgramId = useSelector(({ workingListsContext }) => workingListsContext?.teiList?.programIdView);
```
Comment (henrikmv) [inline src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/hooks/useWorkingListLabel.js:24]: For me it looks like the two useSelector calls are rendered 10 times each whenever navigating to the dashboard for a TEI. Should this behavior be optimized?"

Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Breadcrumbs/EnrollmentBreadcrumb/hooks/useWorkingListLabel.js:24]: It looks like `workingListProgramId` triggers one render and `workingListTemplates` triggers two. (One of each is batched, so they should only really trigger one rerender. Did you have any other experience, @henrikmv? Remember, if you want to see it the value triggers a rerender, you will have to wrap the `console.log` inside a useEffect.

**Comment by eirikhaugstulen** [comment]
Thanks @henrikmv! Moving the issue to testing for now, but we can still continue discussing the rest of the comments 😊

**Review by henrikmv — APPROVED** [review]

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.42,2.41.3,2.40.6,2.39.8 versions

**Pull request merged**